kernel/userspace.c

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+#include <kernel.h>
+#include <string.h>
+#include <misc/printk.h>
+#include <kernel_structs.h>
+#include <sys_io.h>
+
+/**
+ * Kernel object validation function
+ *
+ * Retrieve metadata for a kernel object. This function is implemented in
+ * the gperf script footer, see gen_kobject_list.py
+ *
+ * @param obj Address of kernel object to get metadata
+ * @return Kernel object's metadata, or NULL if the parameter wasn't the
+ * memory address of a kernel object
+ */
+__weak _GENERIC_SECTION(.kobject_data.text)
+struct _k_object *_k_object_find(void *obj)
+{
+	/* This is the __weak stub implementation, the version created by
+	 * gperf will replace it in build phase 2
+	 */
+	ARG_UNUSED(obj);
+	return NULL;
+}
+
+#ifdef CONFIG_PRINTK
```

Comment:
No need for this #if.  The function won't be linked if unreferenced, and it's not static so it won't generate an error.

Reply:
yes, there is.
gc-sections is broken with C strings. even if gc-sections strips this function, the strings will still be in the binary :(
upstream fixed it, but i believe the fix landed in the GCC 7.x branch.